### PR TITLE
Refactor DataCollectorV0 and HDF5 dependencies isolation

### PR DESCRIPTION
**Pull request by younik**
## Motivation

In the future, we aim to consider different storage methods than HDF5 files for our datasets (see https://github.com/Farama-Foundation/Minari/issues/98), as other backends can be faster.
Currently, different files depend on h5py; this PR isolates the dependency on MinariStorage. 
If we add support for other storage, we simply need to create a new MinariStorage that implements the same interface.

## Changes

To achieve the h5py isolation, we make DataCollectorV0, functions in utils and MinariDataset to interface with MinariStorage instead of directly to h5py. Thus, MinariStorage, should offer an API for the most used methods. 
This makes as natural dependencies `DataCollectorV0 -> MinariStorage.` and `MinariDataset -> MinariStorage`.

To avoid `MinariDataset -> DataCollectorV0` dependency, we change the current API for adding the buffer of a DataCollector to a MInariDataset, from:
```python
minari_dataset.update_from_env_collector(data_collector)
```
to:
```python
data_collector.add_to_dataset(minari_dataset)
```

Other user notable changes:
 - MinariStorage doesn't want anymore the path to the file, but the path to the directory containing the file(s).
 - The MinariStorage of a MinariDataset is now accessible using `.storage`
 - The init method of MinariStorage is for an existing dataset. To create a new storage, use the class method `new`
 - It is not possible anymore to combine dataset without copying. We may want to add this again in the future (but it will likely be incompatible with different backends).
  - Remove the custom error when env module is missing. There is already an informative error from Gymnasium (and if it is not informative enough, it should be changed there).

### CODE REVIEW

File: tests/utils/test_dataset_combine.py [hunk 1]
```diff
@@ -173,32 +160,17 @@ def test_combine_datasets():
...
-    # testing with copy
-    combined_dataset = combine_datasets(
-        test_datasets, new_dataset_id="cartpole-combined-test-v0", copy=True
-    )
-    assert combined_dataset.spec.env_spec.max_episode_steps is None
-    _check_load_and_delete_dataset("cartpole-combined-test-v0")
-
     # Check that we get max(max_episode_steps) when there is no max_episode_steps=None
     test_datasets.pop()
-    # testing without creating a copy
```
Comment (balisujohn) [inline tests/utils/test_dataset_combine.py:192]: What is the reason for removing these checks?

File: tests/dataset/test_minari_dataset.py [hunk 2]
```diff
@@ -58,9 +62,9 @@ def test_episode_data(space: gym.Space):
 @pytest.mark.parametrize(
     "dataset_id,env_id",
     [
-        ("cartpole-test-v0", "CartPole-v1"),
-        ("dummy-dict-test-v0", "DummyDictEnv-v0"),
-        ("dummy-box-test-v0", "DummyBoxEnv-v0"),
+        # ("cartpole-test-v0", "CartPole-v1"),
```
Comment (balisujohn) [inline tests/dataset/test_minari_dataset.py:65]: I think these need to be uncommented

File: tests/utils/test_dataset_combine.py (hunk 1 above)
Comment (younik) [reply tests/utils/test_dataset_combine.py:192]: It is not possible anymore to have copy=False, as it was causing problem (the ids of episode were not consistent)

**Review by balisujohn — COMMENTED** [review]
I noticed there is still an `h5py` import in utils.py and in it is used in `get_normalized_scores` Is this by design, and if so, what is blocking removing the h5py dependency there?

File: tests/utils/test_dataset_creation.py [hunk 3]
```diff
@@ -50,15 +50,11 @@ def test_generate_dataset_with_collector_env(dataset_id, env_id):
...
     for episode in range(num_episodes):
-        terminated = False
-        truncated = False
-        while not terminated and not truncated:
+        done = False
+        while not done:
             action = env.action_space.sample()  # User-defined policy function
             _, _, terminated, truncated, _ = env.step(action)
-            if terminated or truncated:
-                assert not env._buffer[-1]
```
Comment (balisujohn) [inline tests/utils/test_dataset_creation.py:59]: why is it safe to remove these asserts?

File: tests/utils/test_dataset_combine.py (hunk 1 above)
Comment (younik) [reply tests/utils/test_dataset_combine.py:192]: It is not specific about h5py isolation, but it is a bug in the current code: when you combine two datasets, the episode ids of the second dataset are modified when you have copy=False, i.e. they don't start anymore from 0

File: tests/utils/test_dataset_creation.py (hunk 3 above)
Comment (younik) [reply tests/utils/test_dataset_creation.py:59]: I changed the internal structure and the assert is not true anymore (when resetted, the buffer is not void, but immediately has the episode id)
Anyway, it wasn't a meaningful test (it doesn't test any wanted feature)

**Comment by younik** [comment]
> 

> I noticed there is still an `h5py` import in utils.py and in it is used in `get_normalized_scores` Is this by design?

Oh, thanks for spotting this, it wasn't by design; I fixed it.

File: tests/common.py [hunk 4]
```diff
@@ -547,7 +549,7 @@ def check_load_and_delete_dataset(dataset_id: str):
 
 
 def create_dummy_dataset_with_collecter_env_helper(
-    dataset_id: str, env: DataCollectorV0, num_episodes: int = 10
+    dataset_id: str, env: DataCollectorV0, num_episodes: int = 10, **kwargs
```
Comment (balisujohn) [inline tests/common.py:552]: what is the purpose of the `kwargs`?

**Review by balisujohn — COMMENTED** [review]
Just had one more question, otherwise, once this is passing pre-commit, I think it's ready to merge :^)

File: tests/common.py (hunk 4 above)
Comment (younik) [reply tests/common.py:552]: It is needed for this new test; they are a general kwarg that can be passed to the`create_dataset_from_collector_env` function  https://github.com/Farama-Foundation/Minari/blob/c5c88e8a03921ac8012a632c741916560670acc2/tests/utils/test_get_normalized_score.py#L20-L27

**Review by balisujohn — APPROVED** [review]
Ready to merge :^) but one last question, what it two datasets that are themselves each the result of combining two datasets are combined? Will this situation be handled correctly?

**Comment by younik** [comment]
> Ready to merge :^) but one last question, what it two datasets that are themselves each the result of combining two datasets are combined? Will this situation be handled correctly?

Thanks! Yes, this should not create any issue as the new generated dataset is a perfectly normal dataset (everything is copied); with just different metadata.

**Pull request merged**